elyra/pipeline/component_parser_kfp.py

```diff
@@ -140,28 +138,25 @@ def _read_component_yaml(self, registry_entry: SimpleNamespace) -> Optional[Dict
                            f"location: '{registry_entry.location}' -> {str(e)}")
             return None
 
-    def _get_adjusted_parameter_fields(self,
-                                       component_body: Dict[str, Any],
-                                       io_object_name: str,
-                                       io_object_type: str,
-                                       parameter_type: str) -> str:
+    def _is_pathbased_parameter(self, parameter_name: str, component_body: Dict[str, Any]) -> bool:
         """
-        Change the parameter ref according if it is a KFP path parameter (as opposed to a value parameter)
+        Check whether parameter is a KFP path parameter (as opposed to a value parameter)
+
+        :param parameter_name: the name of the parameter that will be checked
+        :param component_body: the component YAML contents
         """
-        adjusted_type = parameter_type
-        if "implementation" in component_body and "container" in component_body['implementation']:
-            if "command" in component_body['implementation']['container']:
-                for command in component_body['implementation']['container']['command']:
-                    if isinstance(command, dict) and list(command.values())[0] == io_object_name and \
-                            list(command.keys())[0] == f"{io_object_type}Path":
-                        adjusted_type = f"{io_object_type}Path"
-            if "args" in component_body['implementation']['container']:
-                for arg in component_body['implementation']['container']['args']:
-                    if isinstance(arg, dict) and list(arg.values())[0] == io_object_name and \
-                            list(arg.keys())[0] == f"{io_object_type}Path":
-                        adjusted_type = f"{io_object_type}Path"
-
-        return adjusted_type
+        # Get component_body['implementation']['container'] sub-dictionary if it exists
+        component_impl = component_body.get('implementation', {}).get('container', {})
+
+        # Get list of component commands/arguments and check for given parameter
+        commands_and_args = component_impl.get("command", []) + component_impl.get("args", [])
+        for parameter in [c for c in commands_and_args if isinstance(c, dict)]:
+            if parameter_name in list(parameter.values()):
+                if "Path" in list(parameter.keys())[0]:
```

Comment:
This seems to only consider the first key of the `parameter` dictionary while all values of that dictionary are considered, which I don't find intuitive.  If this is expected perhaps we could add a comment or two?  

I also don't really follow why only dictionaries are considered in L153.

Reply:
Yeah, it's a very weird set of conditions all based on how the component YAML is structured. 

I'll explain each but then add some comments as well.

> I also don't really follow why only dictionaries are considered in L153

The resulting list (if it exists) from `.get('command')` and `.get('args')` consists of mixed types (strings and dicts as far as I've encountered). Here is the `command` portion of the papermill component, for example, that is converted into a dictionary in the parser by `yaml.safe_load()`

```
command:
    - sh
    - -exc
    - |
      input_notebook_path="$0"
      output_notebook_path="$1"
      arguments="$2"
      packages_to_install="$3"
      input_data_path="$4"
      output_data_path="$5"
      mkdir -p "$(dirname "$output_notebook_path")"
      mkdir -p "$output_data_path"

      # Converting packages_to_install from JSON to command-line arguments
      packages_to_install=$(echo "$packages_to_install" | sed -E -e 's/^\[//' -e 's/]$//' -e 's/",/" /g' -e "s/\"/'/g")
      # Installing packages
      sh -c "python3 -m pip install --upgrade --quiet jupyter papermill==2.2.0 ${packages_to_install}"
      # Running the notebook using papermill
      papermill --parameters_yaml "$arguments" --parameters INPUT_DATA_PATH "$input_data_path" --parameters OUTPUT_DATA_PATH "$output_data_path" "$input_notebook_path" "$output_notebook_path"
      
    - {inputPath: Notebook}
    - {outputPath: Notebook}
    - {inputValue: Parameters}
    - if:
        cond: {isPresent: Packages to install}
        then: [{inputValue: Packages to install}]
        else: "{}"
    - if:
        cond: {isPresent: Input data}
        then: [{inputPath: Input data}]
        else: ""
    - {outputPath: Output data}
```

`sh` and `-exc` appear as strings, e.g. whereas the entries corresponding to parameters are always dictionaries as in `{inputPath: Notebook}` . The conditionals (`if`) also appear as strings. The same applies if the component includes an `args` sub-dictionary.

> This seems to only consider the first key of the parameter dictionary while all values of that dictionary are considered, which I don't find intuitive. 

Since we're only interested in dicts of the form `{inputPath: Notebook}`, the resulting list of values provides the full parameter name. As an example, `'Notebook'` is in the list `['Notebook']`. The same doesn't apply to the keys() search, though because `Path` is not in `['inputPath']`, but `Path` is in the string 'inputPath'